src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetDestruction.java

```diff
@@ -348,125 +325,70 @@ public static boolean validBlock(World world, BlockPos voidPos, EntityPlayer pla
-        if (tag == null) return;
-
-        BlockPos startPos = NBTUtil.readBlockPos(tag.getCompound(NBTKeys.GADGET_START_POS));
-        int[] indexPosArray = tag.getIntArray(NBTKeys.MAP_INDEX2POS);
-        int[] indexStateIDArray = tag.getIntArray(NBTKeys.MAP_INDEX2STATE_ID);
-        int[] posPasteArray = tag.getIntArray(NBTKeys.MAP_POS_PASTE);
-        int[] statePasteArray = tag.getIntArray(NBTKeys.MAP_STATE_PASTE);
-
-        BlockMapIntState intState = new BlockMapIntState();
-        intState.getIntStateMapFromNBT((NBTTagList) tag.getTag(NBTKeys.MAP_PALETTE));
-
-        boolean success = false;
-        for (int i = 0; i < indexPosArray.length; i++) {
-            BlockPos placePos = GadgetUtils.relIntToPos(startPos, indexPosArray[i]);
-            IBlockState currentState = world.getBlockState(placePos);
-            if (currentState.getBlock().isAir(currentState, world, placePos) || currentState.getMaterial().isLiquid()) {
-                IBlockState placeState = intState.getStateFromSlot((short) indexStateIDArray[i]);
-                if (placeState.getBlock() == BGBlocks.constructionBlock) {
-                    IBlockState pasteState = Blocks.AIR.getDefaultState();
-                    for (int j = 0; j < posPasteArray.length; j++) {
-                        if (posPasteArray[j] == indexPosArray[i]) {
-                            pasteState = intState.getStateFromSlot((short) statePasteArray[j]);
-                            break;
-                        }
-                    }
-                    if (pasteState != Blocks.AIR.getDefaultState()) {
-                        world.spawnEntity(new BlockBuildEntity(world, placePos, player, pasteState, BlockBuildEntity.Mode.PLACE, true));
-                        success = true;
-                    }
-                } else {
-                    world.spawnEntity(new BlockBuildEntity(world, placePos, player, placeState, BlockBuildEntity.Mode.PLACE, false));
-                    success = true;
-                }
-            }
-        }
-        if (success) {
-            NBTTagCompound newTag = new NBTTagCompound();
-            worldSave.addToMap(getUUID(stack), newTag);
-            worldSave.markForSaving();
-        }
+
+        NBTTagCompound serializedSnapshot = worldSave.getCompoundFromUUID(getUUID(stack));
+        if (serializedSnapshot.isEmpty())
+            return;
+
+        RegionSnapshot snapshot = RegionSnapshot.deserialize(serializedSnapshot);
+        restoreSnapshotWithBuilder(world, snapshot);
+        worldSave.addToMap(getUUID(stack), new NBTTagCompound());
+        worldSave.markDirty();
     }
 
     private boolean destroyBlock(World world, BlockPos voidPos, EntityPlayer player) {
         ItemStack tool = getGadget(player);
         if (tool.isEmpty())
             return false;
 
-        if(!this.canUse(tool, player))
+        if (!this.canUse(tool, player))
+            return false;
+
+        if (world.isAirBlock(voidPos))
```

Comment:
Do this before any other check. Mostly for computation enhancements. Do the 'Nothing should haven if false' things before the 'Can we do this' checks. If you know what I mean.